src/main/java/io/vertx/httpproxy/impl/CachingFilter.java

```diff
@@ -102,49 +95,59 @@ private Future<ProxyResponse> tryHandleProxyRequestFromCache(ProxyContext contex
-      if (cacheControl.maxAge() >= 0) {
-        long now = System.currentTimeMillis();
-        long currentAge = now - resource.timestamp;
-        if (currentAge > cacheControl.maxAge() * 1000) {
-          String etag = resource.headers.get(HttpHeaders.ETAG);
-          if (etag != null) {
-            proxyRequest.headers().set(HttpHeaders.IF_NONE_MATCH, resource.etag);
-            context.set("cached_resource", resource);
+      long now = System.currentTimeMillis();
+      long val = resource.getTimestamp() + resource.getMaxAge();
+      if (val < now) {
+        return cache.remove(cacheKey).compose(v -> context.sendRequest());
+      }
+
+      String cacheControlHeader = response.getHeader(HttpHeaders.CACHE_CONTROL);
+      if (cacheControlHeader != null) {
+        CacheControl cacheControl = new CacheControl().parse(cacheControlHeader);
+        if (cacheControl.maxAge() >= 0) {
+          long currentAge = now - resource.getTimestamp();
+          if (currentAge > cacheControl.maxAge() * 1000) {
+            String etag = resource.getHeaders().get(HttpHeaders.ETAG);
+            if (etag != null) {
+              proxyRequest.headers().set(HttpHeaders.IF_NONE_MATCH, resource.getEtag());
+              context.set("cached_resource", resource);
+            }
             return context.sendRequest();
-          } else {
-            return null;
           }
         }
       }
-    }
 
-    //
-    String ifModifiedSinceHeader = response.getHeader(HttpHeaders.IF_MODIFIED_SINCE);
-    if ((response.method() == HttpMethod.GET || response.method() == HttpMethod.HEAD) && ifModifiedSinceHeader != null && resource.lastModified != null) {
-      Instant ifModifiedSince = ParseUtils.parseHeaderDate(ifModifiedSinceHeader);
-      if (!ifModifiedSince.isAfter(resource.lastModified)) {
-        response.response().setStatusCode(304).end();
-        return Future.succeededFuture();
+      //
+      String ifModifiedSinceHeader = response.getHeader(HttpHeaders.IF_MODIFIED_SINCE);
+      if ((response.method() == HttpMethod.GET || response.method() == HttpMethod.HEAD) && ifModifiedSinceHeader != null && resource.getLastModified() != null) {
+        Instant ifModifiedSince = ParseUtils.parseHeaderDate(ifModifiedSinceHeader);
+        if (!ifModifiedSince.isAfter(resource.getLastModified())) {
+          return Future.succeededFuture(proxyRequest.release().response().setStatusCode(304));
+        }
       }
-    }
-    proxyRequest.release();
-    ProxyResponse proxyResponse = proxyRequest.response();
-    resource.init(proxyResponse);
-    return Future.succeededFuture(proxyResponse);
+      proxyRequest.release();
+      ProxyResponse proxyResponse = proxyRequest.response();
+      fillResponseFromResource(proxyResponse, resource);
+      return Future.succeededFuture(proxyResponse);
+    });
+
+  }
+
+  public void fillResponseFromResource(ProxyResponse proxyResponse, Resource resource) {
+    proxyResponse.setStatusCode(200);
```

Comment:
Why not using the status code of the resource?

Reply:
this part of the code is moved from `Resource#init` where it uses 200 in the original code. In PR I don't want to change any logic so I left it in here, instead I made this change in #93

Reply:
Fair enough